trio/tests/test_sync.py

```diff
@@ -568,3 +568,126 @@ async def lock_taker():
+            wakeups += 1
+
+    async with _core.open_nursery() as nursery:
+        nursery.start_soon(background)
+
+        # The event stream starts in a blocked state (no event fired)
+        await wait_all_tasks_blocked()
+        assert wakeups == 0
+
+        # Calling fire() lets it run:
+        p.fire()
+        await wait_all_tasks_blocked()
+        assert wakeups == 1
+
+        # Multiple events are coalesced into one:
+        p.fire()
+        p.fire()
+        p.fire()
+        await wait_all_tasks_blocked()
+        assert wakeups == 2
+
+        p.close()
+
+
+async def test_EventStream_while_task_is_elsewhere(autojump_clock):
+    p = EventStream()
+
+    wakeups = 0
+
+    async def background():
+        nonlocal wakeups
+        async for _ in p.subscribe():
+            wakeups += 1
+            await sleep(10)
+
+    async with _core.open_nursery() as nursery:
+        nursery.start_soon(background)
+
+        # Double-check that it's all idle and settled waiting for a event
+        await sleep(5)
+        assert wakeups == 0
+        await sleep(10)
+        assert wakeups == 0
+
+        # Wake it up
+        p.fire()
+
+        # Now it's sitting in sleep()...
+        await sleep(5)
+        assert wakeups == 1
+
+        # ...when another event arrives.
+        p.fire()
+
+        # It still wakes up though
+        await sleep(10)
+        assert wakeups == 2
+
+        p.close()
+
+
+async def test_EventStream_subscribe_independence(autojump_clock):
+    p = EventStream()
+
+    wakeups = [0, 0]
+
+    async def background(i, sleep_time):
+        nonlocal wakeups
+        async for _ in p.subscribe():
+            wakeups[i] += 1
+            await sleep(sleep_time)
+
+    try:
+        async with _core.open_nursery() as nursery:
+            nursery.start_soon(background, 0, 10)
+            nursery.start_soon(background, 1, 100)
+
+            # Initially blocked, no event fired
+            await sleep(200)
+            assert wakeups == [0, 0]
+
+            # Firing an event wakes both tasks
+            p.fire()
+            await sleep(5)
+            assert wakeups == [1, 1]
+
+            # Now
+            # task 0 is sleeping for 5 more seconds
+            # task 1 is sleeping for 95 more seconds
+
+            # Fire events at a 10s interval; task 0 will wake up for each
+            # task 1 will only wake up after its sleep
+            p.fire()
+            await sleep(10)
+            p.fire()
+            assert wakeups == [2, 1]
+            await sleep(100)
+            assert wakeups == [3, 2]
+
+            # Now task 0 is blocked on the next event
+            # Task 1 is sleeping for 100s
+
+            p.fire()
+            await sleep(1)
+            assert wakeups == [4, 2]
+            await sleep(100)
+            assert wakeups == [4, 3]
+
+            p.close()
+    except:
+        import traceback
+        traceback.print_exc()
+        raise
```

Comment:
Yes, it was reporting a ResourceWarning about an async generator being disposed:

```
E               Traceback (most recent call last):                                                                                                                       
E                 File "/home/mauve/dev/trio/trio/_core/_asyncgens.py", line 79, in finalizer                                                                           
E                   warnings.warn(                                                                                                                                       
E               ResourceWarning: Async generator 'trio._sync.EventStream.subscribe' was garbage collected before it had been exhausted. Surround its use in 'async with a
closing(...):' to ensure that it gets cleaned up as soon as you're done using it.                                                                                        
```

I didn't dig into what you're doing here, and it sounds a bit odd to me, but I rolled with it. This is one reason I added the `.close()` method (but I think it is a good idea anyway).

Reply:
Huh, ok. A lot going on here :-)

I don't see how the `except` block could interact with that warning. The warning is generated in "GC context", where raising exceptions is impossible. Well, it's really a fake GC context, because we re-enter Trio and run the finalizer in async context, so Trio tries to emulate the interpreter's regular GC context... it's possible we don't do it perfectly and some kind of exception escapes somewhere. But I don't think so?

Also, that warning ought to be harmless in this case. It's pointing out that *if* the async generator is holding any resources like file descriptors, then they are being cleaned up by the GC rather than explicitly. Is this a useful thing to point out? tbh I'm not sure. `ResourceWarning`s are hidden by default so it's not necessarily a big deal to have some false positives. But in this case it's pure noise, since `subscribe` doesn't hold any resources. And CPython doesn't bother pointing this out when it GC's regular sync generators -- the only reason we can do it is because the GC delegates async generator cleanup to Trio. Maybe it would be better to just drop the warning entirely? It's kind of a mess.

@oremanj, any thoughts you want to add?